presto-main/src/main/java/com/facebook/presto/sql/planner/EnforceRequiredJoinMethods.java

```diff
@@ -0,0 +1,209 @@
+import com.facebook.presto.sql.planner.plan.DeleteNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+
+import java.util.Map;
+
+import static com.facebook.presto.SystemSessionProperties.isDistributedJoinEnabled;
+import static com.facebook.presto.sql.planner.optimizations.ScalarQueryUtil.isScalar;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static java.util.Objects.requireNonNull;
+
+public class EnforceRequiredJoinMethods
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(session), plan, new Context(false));
+    }
+
+    private static class Context
+    {
+        private final boolean isDistributedSemiJoinBanned;
+
+        private Context(boolean isDistributedSemiJoinBanned)
+        {
+            this.isDistributedSemiJoinBanned = isDistributedSemiJoinBanned;
+        }
+
+        public boolean isDistributedSemiJoinBanned()
+        {
+            return isDistributedSemiJoinBanned;
+        }
+
+        public Context withDistributedSemiJoinBanned()
+        {
+            return new Context(true);
+        }
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Context>
+    {
+        private final Session session;
+
+        public Rewriter(Session session)
+        {
+            this.session = session;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Context> context)
+        {
+            PlanNode leftRewritten = context.rewrite(node.getLeft(), context.get());
+            PlanNode rightRewritten = context.rewrite(node.getRight(), context.get());
+            boolean distribute = mustDistributeJoin(node);
+            if (distribute && node.getMethod() != JoinNode.Method.DISTRIBUTED) {
+                return new JoinNode(
+                        node.getId(),
+                        node.getType(),
+                        JoinNode.Method.DISTRIBUTED,
+                        leftRewritten,
+                        rightRewritten,
+                        node.getCriteria(),
+                        node.getFilter(),
+                        node.getLeftHashSymbol(),
+                        node.getRightHashSymbol());
+            }
+
+            if (!distribute && node.getMethod() != JoinNode.Method.BROADCAST) {
```

Comment:
I don't think that optimizing single object (linear to number of joins) per query creation is worth 20 lines of code. I would just:
```
if (distrbiubted) {
  return new JoinNode(...);
} else {
  return new JoinNode(...);
}
```